File: python/glasshouse/classification.py
"""Binary classification metrics, weighted. Three kinds, and it matters which you read.

- **Decision** metrics (accuracy, precision, recall, F1, MCC) score a hard yes/no at a
  threshold. They answer "how good is this policy" — not "how good is this model".
- **Ranking** metrics (ROC-AUC, average precision, KS) score the ordering of the scores, so
  they are threshold-free but blind to whether the probabilities are right.
- **Probability** metrics (log-loss, Brier) are proper scores: you cannot game them by
  miscalibrating. Read them with :func:`glasshouse.metrics.calibration_table`.

Labels must be exactly 0/1; ``score`` is a probability or any monotone score (for the
probability metrics it must be a probability in (0, 1)).
"""

from __future__ import annotations

from dataclasses import dataclass

from glasshouse import _core
from glasshouse.metrics import ArrayLike, _f64, _weights, deviance


@dataclass(frozen=True)
class ThresholdMetrics:
    """Everything you can read off the confusion counts at one threshold."""

    tp: float
    fp: float
    fn: float
    tn: float
    accuracy: float
    balanced_accuracy: float
    precision: float
    recall: float
    f1: float
    mcc: float


def threshold_metrics(
    y: ArrayLike, score: ArrayLike, sample_weight: ArrayLike | None = None, threshold: float = 0.5
) -> ThresholdMetrics:
    """Compute the confusion counts and every decision metric at ``score >= threshold``.

    What it is for: judging a *decision rule*, which is what actually gets deployed. Read
    ``mcc`` first under imbalance: it is high only when all four cells are right, so it
    punishes both missing the rare class and crying wolf (Chicco & Jurman 2020). ``f1``
    ignores true negatives; ``accuracy`` is meaningless at 0.2 % positives.

    When it lies: it scores one threshold, not the model — a threshold tuned on the test set
    flatters everything here. Undefined ratios (nothing flagged, one class only) return 0,
    scikit-learn's convention, rather than NaN.

    Examples
    --------
    >>> from glasshouse.classification import threshold_metrics
    >>> m = threshold_metrics([1, 1, 0, 1, 0, 0], [0.9, 0.8, 0.7, 0.4, 0.3, 0.1])
    >>> (m.tp, m.fp, m.fn, m.tn)
    (2.0, 1.0, 1.0, 2.0)
    >>> round(m.mcc, 4)
    0.3333
    """
    w = _weights(sample_weight)
    raw = _core.threshold_metrics(_f64(y, "y"), _f64(score, "score"), w, threshold)
    return ThresholdMetrics(**raw)


def mcc(
    y: ArrayLike, score: ArrayLike, sample_weight: ArrayLike | None = None, threshold: float = 0.5
) -> float:
    """Matthews correlation (phi) at ``threshold``. See :func:`threshold_metrics`."""
    return threshold_metrics(y, score, sample_weight, threshold).mcc


def f1(
    y: ArrayLike, score: ArrayLike, sample_weight: ArrayLike | None = None, threshold: float = 0.5
) -> float:
    """F1 at ``threshold``. See :func:`threshold_metrics`."""
    return threshold_metrics(y, score, sample_weight, threshold).f1


def roc_auc(y: ArrayLike, score: ArrayLike, sample_weight: ArrayLike | None = None) -> float:
    """Area under the ROC curve: ranking quality across all thresholds. 0.5 is random.

    What it is for: balanced classes, comparing models broadly; the probability that a random
    positive outscores a random negative. Ties count half, so it is order-free.

    When it lies: under heavy imbalance a 0.95 can hide a useless precision, because the
    false-positive *rate* stays tiny when negatives are plentiful — read
    :func:`average_precision` there.

    Examples
    --------
    >>> from glasshouse.classification import roc_auc
    >>> roc_auc([0, 0, 1, 1], [0.1, 0.4, 0.35, 0.8])
    0.75
    """
    w = _weights(sample_weight)
    return float(_core.roc_auc(_f64(y, "y"), _f64(score, "score"), w))


def average_precision(
    y: ArrayLike, score: ArrayLike, sample_weight: ArrayLike | None = None
) -> float:
    """Area under the precision-recall curve, step-wise like scikit-learn.

    What it is for: fraud and rare events, where positives are 0.2 % and ROC-AUC looks great
    for everyone. It only looks at how the positives are ranked, so it is the honest ranking
    number under imbalance.

    When it lies: it ignores true negatives entirely, so when both classes matter equally it
    undersells a model; and its baseline moves with prevalence, so never compare it across
    datasets without saying what the positive rate was.

    Examples
    --------
    >>> from glasshouse.classification import average_precision
    >>> round(average_precision([0, 0, 1, 1], [0.1, 0.4, 0.35, 0.8]), 4)
    0.8333
    """
    w = _weights(sample_weight)
    return float(_core.average_precision(_f64(y, "y"), _f64(score, "score"), w))


def ks(y: ArrayLike, score: ArrayLike, sample_weight: ArrayLike | None = None) -> float:
    """Kolmogorov-Smirnov statistic: largest gap between the two classes' score distributions.

    What it is for: credit scoring's favourite; 0 is no separation, 1 is perfect, and it
    names the single score cut where the classes are most different.

    When it lies: it is one point on the curve, so two very different models can tie, and it
    is as blind to calibration as every other ranking metric.

    Examples
    --------
    >>> from glasshouse.classification import ks
    >>> ks([0, 0, 1, 1], [0.1, 0.4, 0.35, 0.8])
    0.5
    """
    w = _weights(sample_weight)
    return float(_core.ks(_f64(y, "y"), _f64(score, "score"), w))


def log_loss(y: ArrayLike, prob: ArrayLike, sample_weight: ArrayLike | None = None) -> float:
    """Mean negative log-likelihood of the labels under ``prob``: half the binomial deviance.

    What it is for: scoring the probabilities themselves. It is a proper score, so the only
    way to improve it is to be more right — you cannot game it by shifting a threshold.

    When it lies: a handful of confident mistakes dominate it (``-ln(1e-6)`` is 13.8), and it
    is hard to explain to a business owner. Pair it with :func:`brier`.

    Examples
    --------
    >>> from glasshouse.classification import log_loss
    >>> round(log_loss([0, 1], [0.2, 0.8]), 4)
    0.2231
    """
    return deviance(y, prob, family="binomial", sample_weight=sample_weight) / 2.0


def brier(y: ArrayLike, prob: ArrayLike, sample_weight: ArrayLike | None = None) -> float:
    """Brier score: mean squared error between labels and probabilities. 0 is perfect.

    What it is for: a proper score you can explain — "on average the probability was off by
    the square root of this". Decomposes into calibration and refinement.

    When it lies: under heavy imbalance the always-say-0 model scores the positive rate, which
    looks excellent; read it against that baseline, not against 0.

    Examples
    --------
    >>> from glasshouse.classification import brier
    >>> round(brier([0, 1], [0.2, 0.8]), 4)
    0.04
    """
    return deviance(y, prob, family="gaussian", sample_weight=sample_weight)


__all__ = [
    "ThresholdMetrics",
    "average_precision",
    "brier",
    "f1",
    "ks",
    "log_loss",
    "mcc",
    "roc_auc",
    "threshold_metrics",
]

File: crates/core/src/classification.rs
//! Binary classification metrics. Weighted, and split by what they score:
//! a *decision* (threshold metrics from the confusion counts), a *ranking* (ROC-AUC, average
//! precision, KS), or the *probabilities themselves* (log-loss and Brier, which are the
//! binomial and gaussian deviances and live in `metrics`).

use crate::error::{all_values, same_length, GlassError};
use crate::ranking::sorted_tie_groups;

/// Weighted confusion counts at a threshold.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Confusion {
    pub tp: f64,
    pub fp: f64,
    pub fn_: f64,
    pub tn: f64,
}

impl Confusion {
    /// Count (with weights) how `score >= threshold` decisions land against 0/1 labels.
    ///
    /// # Errors
    /// Lengths differ; labels not exactly 0 or 1; non-finite scores; bad weights.
    pub fn at(
        y: &[f64],
        score: &[f64],
        w: Option<&[f64]>,
        threshold: f64,
    ) -> Result<Self, GlassError> {
        validate_labels(y, score, w)?;
        let mut c = Self {
            tp: 0.0,
            fp: 0.0,
            fn_: 0.0,
            tn: 0.0,
        };
        for (i, (&yi, &si)) in y.iter().zip(score).enumerate() {
            let wi = w.map_or(1.0, |w| w[i]);
            match (is_positive(yi), si >= threshold) {
                (true, true) => c.tp += wi,
                (false, true) => c.fp += wi,
                (true, false) => c.fn_ += wi,
                (false, false) => c.tn += wi,
            }
        }
        Ok(c)
    }

    #[must_use]
    pub fn accuracy(&self) -> f64 {
        (self.tp + self.tn) / (self.tp + self.tn + self.fp + self.fn_)
    }

    /// Mean of recall on each class. 0.5 is guessing, whatever the class balance.
    #[must_use]
    pub fn balanced_accuracy(&self) -> f64 {
        self.recall()
            .midpoint(zero_if_nan(self.tn / (self.tn + self.fp)))
    }

    /// Of everything flagged, how much was real. 0 when nothing was flagged (scikit-learn's
    /// `zero_division=0` convention).
    #[must_use]
    pub fn precision(&self) -> f64 {
        zero_if_nan(self.tp / (self.tp + self.fp))
    }

    /// Of everything real, how much was flagged.
    #[must_use]
    pub fn recall(&self) -> f64 {
        zero_if_nan(self.tp / (self.tp + self.fn_))
    }

    /// Harmonic mean of precision and recall.
    #[must_use]
    pub fn f1(&self) -> f64 {
        zero_if_nan(2.0 * self.tp / (2.0 * self.tp + self.fp + self.fn_))
    }

    /// Matthews correlation (phi): high only when all four cells are right. 0 when any
    /// marginal is empty (undefined; scikit-learn returns 0 there too).
    #[must_use]
    pub fn mcc(&self) -> f64 {
        let Self { tp, fp, fn_, tn } = *self;
        let denom = ((tp + fp) * (tp + fn_) * (tn + fp) * (tn + fn_)).sqrt();
        zero_if_nan((tp * tn - fp * fn_) / denom).clamp(-1.0, 1.0)
    }
}

/// Labels are validated to be exactly 0 or 1, so exact comparison is the semantics here.
#[allow(clippy::float_cmp)]
#[inline]
fn is_positive(label: f64) -> bool {
    label == 1.0
}

#[allow(clippy::float_cmp)]
#[inline]
fn is_label(v: f64) -> bool {
    v == 0.0 || v == 1.0
}

/// Bounded metrics can land a rounding error outside [0, 1] (seen on macOS CI:
/// 1.0000000000000002); the true value cannot, so clamp to the real range.
#[inline]
fn unit(v: f64) -> f64 {
    v.clamp(0.0, 1.0)
}

fn zero_if_nan(v: f64) -> f64 {
    if v.is_nan() {
        0.0
    } else {
        v
    }
}

/// Area under the ROC curve, weighted, ties as one step (trapezoid). 0.5 is random.
///
/// # Errors
/// As [`Confusion::at`], plus both classes must be present.
pub fn roc_auc(y: &[f64], score: &[f64], w: Option<&[f64]>) -> Result<f64, GlassError> {
    let (pos, neg) = validate_ranking(y, score, w)?;
    let weight = |row: usize| w.map_or(1.0, |w| w[row]);
    let (order, groups) = sorted_tie_groups(score);
    // walk from the highest score down: each tie group moves (fpr, tpr) by one step
    let (mut tp, mut fp, mut area) = (0.0, 0.0, 0.0);
    for &(start, end) in groups.iter().rev() {
        let (mut dtp, mut dfp) = (0.0, 0.0);
        for &row in &order[start..end] {
            if is_positive(y[row]) {
                dtp += weight(row);
            } else {
                dfp += weight(row);
            }
        }
        let (tpr0, fpr0) = (tp / pos, fp / neg);
        tp += dtp;
        fp += dfp;
        let (tpr1, fpr1) = (tp / pos, fp / neg);
        area += (fpr1 - fpr0) * (tpr0 + tpr1) / 2.0;
    }
    Ok(unit(area))
}

/// Average precision: the area under the precision–recall curve as a step function, the same
/// definition as scikit-learn's `average_precision_score`. The baseline is the positive rate.
///
/// # Errors
/// As [`roc_auc`].
pub fn average_precision(y: &[f64], score: &[f64], w: Option<&[f64]>) -> Result<f64, GlassError> {
    let (pos, _) = validate_ranking(y, score, w)?;
    let weight = |row: usize| w.map_or(1.0, |w| w[row]);
    let (order, groups) = sorted_tie_groups(score);
    let (mut tp, mut fp, mut ap, mut prev_recall) = (0.0, 0.0, 0.0, 0.0);
    for &(start, end) in groups.iter().rev() {
        for &row in &order[start..end] {
            if is_positive(y[row]) {
                tp += weight(row);
            } else {
                fp += weight(row);
            }
        }
        let recall = tp / pos;
        let precision = tp / (tp + fp);
        ap += (recall - prev_recall) * precision;
        prev_recall = recall;
    }
    Ok(unit(ap))
}

/// Kolmogorov–Smirnov: the largest gap between the score distributions of the two classes.
/// 0 is no separation, 1 is perfect.
///
/// # Errors
/// As [`roc_auc`].
pub fn ks(y: &[f64], score: &[f64], w: Option<&[f64]>) -> Result<f64, GlassError> {
    let (pos, neg) = validate_ranking(y, score, w)?;
    let weight = |row: usize| w.map_or(1.0, |w| w[row]);
    let (order, groups) = sorted_tie_groups(score);
    let (mut cum_pos, mut cum_neg, mut best) = (0.0, 0.0, 0.0_f64);
    for (start, end) in groups {
        for &row in &order[start..end] {
            if is_positive(y[row]) {
                cum_pos += weight(row);
            } else {
                cum_neg += weight(row);
            }
        }
        best = best.max((cum_pos / pos - cum_neg / neg).abs());
    }
    Ok(unit(best))
}

fn validate_labels(y: &[f64], score: &[f64], w: Option<&[f64]>) -> Result<(), GlassError> {
    if y.is_empty() {
        return Err(GlassError::Empty { name: "y" });
    }
    same_length("y", y, "score", score)?;
    all_values(
        "y",
        y,
        "must be exactly 0 or 1",
        "binary metrics need hard labels; encode the positive class as 1",
        is_label,
    )?;
    all_values(
        "score",
        score,
        "must be finite",
        "NaN or inf cannot be scored",
        f64::is_finite,
    )?;
    if let Some(w) = w {
        same_length("y", y, "sample_weight", w)?;
        all_values(
            "sample_weight",
            w,
            "must be finite and >= 0",
            "negative or NaN weights are not weights",
            |v| v.is_finite() && v >= 0.0,
        )?;
    }
    Ok(())
}

/// Labels + both classes present (weighted). Returns (total positive weight, total negative).
fn validate_ranking(y: &[f64], score: &[f64], w: Option<&[f64]>) -> Result<(f64, f64), GlassError> {
    validate_labels(y, score, w)?;
    let weight = |row: usize| w.map_or(1.0, |w| w[row]);
    let pos: f64 = y
        .iter()
        .enumerate()
        .filter(|(_, &v)| is_positive(v))
        .map(|(i, _)| weight(i))
        .sum();
    let neg: f64 = y
        .iter()
        .enumerate()
        .filter(|(_, &v)| !is_positive(v))
        .map(|(i, _)| weight(i))
        .sum();
    if pos <= 0.0 || neg <= 0.0 {
        return Err(GlassError::InvalidValues {
            name: "y",
            count: y.len(),
            rule: "must contain both classes (with positive weight)",
            fix: "a ranking metric needs positives and negatives to separate",
        });
    }
    Ok((pos, neg))
}

#[cfg(test)]
mod tests {
    use super::*;

    const Y: [f64; 6] = [1.0, 1.0, 0.0, 1.0, 0.0, 0.0];
    const S: [f64; 6] = [0.9, 0.8, 0.7, 0.4, 0.3, 0.1];

    #[test]
    fn confusion_at_half() {
        let c = Confusion::at(&Y, &S, None, 0.5).unwrap();
        assert_eq!((c.tp, c.fp, c.fn_, c.tn), (2.0, 1.0, 1.0, 2.0));
        assert!((c.accuracy() - 4.0 / 6.0).abs() < 1e-15);
        assert!((c.precision() - 2.0 / 3.0).abs() < 1e-15);
        assert!((c.recall() - 2.0 / 3.0).abs() < 1e-15);
        assert!((c.f1() - 2.0 / 3.0).abs() < 1e-15);
        assert!((c.mcc() - 1.0 / 3.0).abs() < 1e-15);
    }

    #[test]
    fn perfect_and_random_ranking() {
        assert!((roc_auc(&[0.0, 1.0], &[0.1, 0.9], None).unwrap() - 1.0).abs() < 1e-15);
        assert!((roc_auc(&[0.0, 1.0], &[0.5, 0.5], None).unwrap() - 0.5).abs() < 1e-15);
        assert!((average_precision(&[0.0, 1.0], &[0.1, 0.9], None).unwrap() - 1.0).abs() < 1e-15);
        assert!((ks(&[0.0, 1.0], &[0.1, 0.9], None).unwrap() - 1.0).abs() < 1e-15);
        assert!(ks(&[0.0, 1.0], &[0.5, 0.5], None).unwrap().abs() < 1e-15);
    }

    #[test]
    #[allow(clippy::float_cmp)] // asserting the exact 0.0 convention
    fn undefined_mcc_is_zero_like_sklearn() {
        let c = Confusion::at(&[1.0, 1.0], &[0.9, 0.8], None, 0.5).unwrap();
        assert_eq!(c.mcc(), 0.0);
    }

    #[test]
    fn refuses_soft_labels_and_one_class() {
        assert!(Confusion::at(&[0.5, 1.0], &[0.1, 0.9], None, 0.5).is_err());
        assert!(roc_auc(&[1.0, 1.0], &[0.1, 0.9], None).is_err());
    }
}
